pkg/bpf: fix bug where second value can be skipped.

tommyp1ckles opened the pull request:
DumpReliablyWithCallback will skip a value callback in some situations.

In cases where the initial key is deleted just after being retrieved, there is no previous key to fallback on. The reliable dump will attempt to use the current nextKey (that was based on the deleted current key).

The local currentKey and nextKey Key types are being passed to NextKey which eventually writes the nextKey output to the nextKeys pointer location (via the bpf syscall).

The currentValue was simply being assinged by the equals operator, which was copying the underlying interface pointer.

Thus in this situation, the next iteration attempt was passing the same pointer twice to NextKey causing the currentKey to be set to the next key a second time - skipping a map element.

Fixes #26491

## Validating the fix

Because the test has [256 possible keys](https://github.com/cilium/cilium/blob/bd9c7cdcc8fe10ee6eb225992dee985b213a9367/pkg/bpf/map_linux_test.go#L505) the particular start condition for this failure seems to be quite rare in CI, by reducing this to 5 it happens much more often (presumably 20% of the time?) - making this much easier to reproduce.

I tested this change by continuously running the DumpReliablyWithCallback (modified with maxSize=5) test to reproduce, and after the fix to validate (ran it 1000 times post fix without failure).

```release-note
Fix bug where bpf map entries may not be reliably dumped or garbage collected when the map is actively being updated.
```

tommyp1ckles committed:
pkg/bpf: fix bug where bpf map entries may not be reliably dumped.

DumpReliablyWithCallback will skip a value callback in some situations.
This may result in incorrect cilium map dumps or garbage collection.

In situations where the initial key is deleted just after being retrieved, there is no previous key to fallback on.
The reliable dump will attempt to use the current nextKey (that was based on the deleted current
key).

The local currentKey and nextKey Key types are being passed to NextKey
which eventually writes the nextKey output to the nextKeys pointer
location (via the bpf syscall).

The currentValue was simply being assinged by the equals operator, which
was copying the underlying interface pointer.

Thus in this situation, the next iteration attempt was passing the same
pointer twice to NextKey causing the currentKey to be set to the next
key a second time - skipping a map element.

Fixes #26491

Signed-off-by: Tom Hadlaw <tom.hadlaw@isovalent.com>

rgo3 commented on pkg/bpf/map_linux.go:
```diff
@@ -671,7 +671,11 @@ func (m *Map) DumpReliablyWithCallback(cb DumpCallback, stats *DumpStats) error
+				// We can't simply assign currentKey to equal nextKey because
+				// that will result in currentKey and nextKey pointing to the
+				// same memory (which causes unexpected behavior when calling ebpf.NextKey).
+				m.NextKey(currentKey, currentKey)
```
What's the behavior of `m.NextKey` if you call it with a key that got deleted? Will it set `currentKey` to the first key in the map? And if so couldn't we just do `m.NextKey(nil, currentKey)`? I think the comment could explain this a bit more, in addition to explaining why `currentKey = nextKey` isn't working.

And shouldn't we check the returned error?

tommyp1ckles replied:
Looking at this again, I think I was overthinking this.  I have a simpler solution that avoids the extra NetKey call now.

tommyp1ckles committed:
pkg/bpf: add additional unit test to avoid dump bugs/regressions.

This test is an improvement on the previous TestDumpReliablyWithCallback test.
The goal of this one is to provide more robust testing of the reliable dump mechanism.
Specifically, it does the following:

 1. Creates a map with a small number of entries, populate it with [1, maxEntries)
 2. Start a goroutine that continuously dumps the map and checks that the dump contains
    all odd elements in the range [1, maxEntries).
 3. At the same time, start another goroutine that continuously deletes and re-adds even
    elements in the map.

The motivation here is to provide a test that better catches regressions in code that is inherently prone to race-condition.

This creates a situation where we have interleaved updates and dumps, we're interested
in ensuring that each dump contains all odd elements in the range [1, maxEntries).
This will catch bugs and regressions related where elements are skipped in the dump.

For example, while running this without the fix: 74f71841e9c037ddd10bedc3128f3b28cb023597 this will fail a majority of the time.
Following this fix, this test should always pass.

This was tested locally by running this several times with a million iterations - both with the fix and without.
For practical purposes we will lower the number of iterations to 1000 to avoid slowing down the test suite too much.

Signed-off-by: Tom Hadlaw <tom.hadlaw@isovalent.com>

rgo3 commented on pkg/bpf/map_linux_test.go:
```diff
@@ -501,6 +502,97 @@ func (s *BPFPrivilegedTestSuite) TestDump(c *C) {
+	// even if the even keys are being deleted and reinserted.
+	expect := map[string]string{}
+	for i := uint32(0); i < maxEntries; i++ {
+		if i != 0 && i%2 != 0 {
```
Nit, 0%2 == 0:
```suggestion
		if i%2 != 0 {
```

rgo3 commented on pkg/bpf/map_linux_test.go:
```diff
@@ -501,6 +502,97 @@ func (s *BPFPrivilegedTestSuite) TestDump(c *C) {
+		ds := NewDumpStats(m)
+		err := m.DumpReliablyWithCallback(func(key MapKey, value MapValue) {
+			k := key.(*TestKey).Key
+			if k != 0 && k%2 != 0 {
```
```suggestion
			if k%2 != 0 {
```

rgo3 commented on pkg/bpf/map_linux_test.go:
```diff
@@ -501,6 +502,97 @@ func (s *BPFPrivilegedTestSuite) TestDump(c *C) {
+			default:
+			}
+
+			for i := uint32(2); i < maxEntries; i += 2 {
```
Why aren't we starting at 0? Even if the first key is deleted/reinserted it should work right?

tommyp1ckles replied:
correct - i'll change it to 0.

rgo3 commented on pkg/bpf/map_linux_test.go:
```diff
@@ -501,6 +502,97 @@ func (s *BPFPrivilegedTestSuite) TestDump(c *C) {
+				dump[fmt.Sprintf("key=%d", key.(*TestKey).Key)] = fmt.Sprintf("value=%d", value.(*TestValue).Value)
+			}
+		}, ds)
+		if err == nil {
```
Should the test fail if `err != nil`?

tommyp1ckles replied:
No, the reliable dump has a max number of iterations it can fall back to previous keys before failing, however as with the other test I'll add the assertion:
```go
c.Check(err, Equals, ErrMaxLookup)
```

If an error is returned - which is the correct behavior.

rgo3 reviewed (changes requested):
Just a couple of nits/comments, otherwise LGTM.
